### PR TITLE
[pallas] Document the indexing_mode=Unblocked()

### PR DESCRIPTION
**Pull request by gnecula**
In the process discovered that the padding in the interpreter mode was with 0s. I changed it to NaN/minint to match the padding for the blocked mode.

### CODE REVIEW

File: docs/pallas/grid_blockspec.md [hunk 1]
```diff
@@ -107,7 +103,10 @@ to get the actual element index on the corresponding array axis.
 If the block shape does not divide evenly the overall shape then the
 last iteration on each axis will still receive references to blocks
 of `block_shape` but the elements that are out-of-bounds are padded
-on input and discarded on output. Note that at least one of the
+on input and discarded on output. The padding is unspecified, and
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:106]: "The values in the padding are unspecified" to be a little more concrete? "padding is unspecified" sounds like the amount of padding is undetermined which is not true

File: docs/pallas/grid_blockspec.md [hunk 2]
```diff
@@ -107,7 +103,10 @@ to get the actual element index on the corresponding array axis.
...
+you should assume it is garbage. In the `interpret=True` mode, we
+pad with NaN for floating-point values, to give users a chance to
+spot accessing out-of-bounds elements. Note that at least one of the
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:109]: ... elements, _but this behavior should not be depended upon_.

File: docs/pallas/grid_blockspec.md [hunk 3]
```diff
@@ -124,7 +123,7 @@ below:
 ...   block_indices = x_spec.index_map(*invocation_indices)
 ...   assert len(x_shape) == len(x_spec.block_shape) == len(block_indices)
 ...   elem_indices = []
-...   for x_size, block_size, block_idx in zip(x_shape, x_spec.block_shape, block_indices):
+...   for x_size, block_size, block_idx in zip(x_shape, x_spec.block_shape, block_indices):
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:126]: unnecessary whitespace

File: docs/pallas/grid_blockspec.md [hunk 4]
```diff
@@ -233,3 +234,44 @@ See [the Pallas TPU documentation](https://jax.readthedocs.io/en/latest/pallas/t
  [309 309 309 319 319 319]]
 
 ```
+
+The behavior documented above applies to the `indexing_mode=pl.Blocked()`.
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:238]: Should we have section headers to make it clearer that we transition to a different mode?

File: docs/pallas/grid_blockspec.md [hunk 5]
```diff
@@ -233,3 +234,44 @@ See [the Pallas TPU documentation](https://jax.readthedocs.io/en/latest/pallas/t
...
+When using the `pl.Unblocked` indexing mode the values returned by the
+index map function are used directly as the array indices, without first
+scaling them by the block size.
+When using the unblocked mode you can specify virtual padding
+of the array as a tuple of low-high paddings for each dimension: the
+behavior is as if the overall array is padded on input. No guarantees
+are made for the padding in the unblocked mode, similarly to the padding
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:245]: "for the padding values"

**Review by apaszke — APPROVED** [review]

File: docs/pallas/grid_blockspec.md (hunk 4 above)
Comment (gnecula) [reply docs/pallas/grid_blockspec.md:238]: I added a section header for this. We should revisit the structure if and when we are adding more material here.
